### ISSUE TITLE
Don't follow imports within 'if TYPE_CHECKING' clauses

### ISSUE THREAD

**Issue by henzef** [pyinstaller/pyinstaller#8860]
**Is your feature request related to a problem? Please describe.**
I use a shared library in my application that has some optional dependencies that are not imported directly, but only used in type annotations in a 'if TYPE_CHECKING:' clause. In other words: The library doesn't use the optional dependencies unless an object of a type from this optional dependency is passed to it.
Unfortunately PyInstaller seems to follow these imports and adds this optional dependency to the generated executable, which is not necessary and bloats the executable.

**Describe the solution you'd like**
PyInstaller should treat TYPE_CHECKING as False and therefore never follow these imports.

**Describe alternatives you've considered**
For now the workaround is to remove the type annotation from the shared library :-(

**Comment by bwoodsend** [comment]
This seems less impossible that I'd expect since the module graph already looks for imports inside conditionals. In [here](https://github.com/pyinstaller/pyinstaller/blob/0779daae400daa4066f91dd634cc731608b901ed/PyInstaller/lib/modulegraph/modulegraph.py#L839-L841), we can detect `if typing.TYPE_CHECKING` via:

```python
isinstance(node.test, ast.Attribute) and node.test.value.id == "typing" and node.test.attr == "TYPE_CHECKING"
```
and `if TYPE_CHECKING` using:
```python
isinstance(node.test, ast.Name) and node.test.id == "TYPE_CHECKING"
```
(The usual caveats about `import tying as alias` or `alias = typing.TYPE_CHECKING` apply.)

But this will surely never amount to more than a mere `--exclude-module=typing_extensions` which, given that `typing_extensions` is so tiny, is only a ~50kB saving (after PYC's zip compression). To me, that doesn't even justify the extra CI/CD it'll take to keep this feature working.

**Comment by rokm** [comment]
I think I'd prefer if users that want to avoid collecting a package pulled in under the `if typing.TYPE_CHECKING` conditional did so using the existing exclusion mechanism - i.e., either use [`--exclude-module` command-line option](https://pyinstaller.org/en/stable/usage.html#cmdoption-exclude-module) (or equivalent `excludes` argument to `Analysis` constructor in the .spec file) or write a custom hook for their package that sets [`excludedimports` hook variable](https://pyinstaller.org/en/stable/hooks.html#hook-global-variables). 

I'm not really opposed to having additional logic to handle `typing.TYPE_CHECKING` and `TYPE_CHECKING` in the modulegraph, but I can imagine situations where this conditional import is the only reference to the package, and automatically removing it will require a hidden import to be added. Plus, relying on modulegraph's conditionals analysis might limit our options to switch to something else in the future (i.e., something less complex).

So all in all, it seems more hassle than its worth, especially since the same effect can be achieved using existing exclusion mechanism.

**Comment by henzef** [comment]
@bwoodsend, the typing_extensions module is not what I meant and it would indeed be neglectable. In my case a single import in this 'if TYPE_CHECKING' results in an increase from 19 to 55 dependant packages due to pulling in a whole dependency tree.

Just consider the following minimal example:

    # my_shared_lib.py
    from typing import Union, TYPE_CHECKING
    if TYPE_CHECKING:
        import numpy as np

    def calculate_something(a: Union[list[int], "np.array"]) -> float:
        ...  # some implementation of a function that *can* handle numpy arrays, but works on other types as well
and

    # my_application.py
    from my_shared_lib import calculate_something
    print(calculate_something([1,2,3]))
and optionally also

    # other_script_that_is_not_in_scope_of_pyinstaller.py
    import numpy as np
    from my_shared_lib import calculate_something
    print(calculate_something(np.array([1, 2, 3]))

In order to compile my_application.py with PyInstaller, numpy would not be needed, but it is pulled in regardless.

**Comment by bwoodsend** [comment]
Fair enough. <sup>(Oh, how I hate typing... it finds ways to create problems even for those that don't use it...)</sup>

I vaguely remember an intention to also exclude imports under `if __name__ == "__main__"` blocks for non-entrypoint scripts – we could probably do both in one go (assuming we do either).

**Comment by henzef** [comment]
Regarding @rokm's comment: To be able to use the existing exclusion mechanisms, the user would need to *know* that the excluded module is indeed only used in 'if TYPE_CHECKING' clauses, which I would consider unrealistic for external modules. Even if I went through the hassle to check this and then add it to pyinstallers exclude list, any update of the external library may start to require the excluded package during runtime and therefore break the generated executable.

**Comment by rokm** [comment]
> Regarding @rokm's comment: To be able to use the existing exclusion mechanisms, the user would need to _know_ that the excluded module is indeed only used in 'if TYPE_CHECKING' clauses, which I would consider unrealistic for external modules. Even if I went through the hassle to check this and then add it to pyinstallers exclude list, any update of the external library may start to require the excluded package during runtime and therefore break the generated executable.

That's the way all conditional imports are handled at the moment, though. Ultimately only *you* know if you really need them or not; and if not, exclude them or ensure they are not installed in the first place (i.e., build in the minimal build environment).

But I agree that typing is a bit of a special case here, and we could try to handle it behind-the-scenes.

**Comment by bwoodsend** [comment]
We really don't want anyone to be building applications without already knowing what should go in it. I know PyInstaller's automatic dependency detection (which predates pip and any form of formal packaging) insinuates otherwise but you should always at least be aware of your top level dependencies and be able to use that list to create minimal venvs whenever you want to avoid bloat.
